### PR TITLE
chore: Prefetch module apis in service worker

### PR DESCRIPTION
**Pull request by dvj1988**
## Description
This PR has the following changes
- Modify the prefetch api cache strategy to handle multiple prefetch requests.
- Convert the service worker files from js to ts
- Code splitting of CE and EE for service worker utils to handle EE specific changes
- Jest test cases for the updated logic

Fixes #`Issue Number`  
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/9511154598>
> Commit: 785197e27873e71ed457f785a73d4ea57f379213
> Cypress dashboard url: <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=9511154598&attempt=1" target="_blank">Click here!</a>

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
  - Enhanced API request handling with new utility functions and caching strategies for service worker operations.
  - Updated service worker configuration to TypeScript for improved type safety and maintainability.
  - Added type definitions for `node-fetch` to support new service worker functionalities.

- **Refactor**
  - Consolidated `view` and `edit` endpoint URL construction in `ConsolidatedPageLoadApi` for better code maintainability.

- **Chores**
  - Updated dependencies in `package.json` for better development experience and compatibility.
  - Added tests for new service worker utility functions to ensure reliability and correctness.

### CODE REVIEW

File: app/client/src/utils/serviceWorkerUtils.test.js
Comment (dvj1988) [inline app/client/src/utils/serviceWorkerUtils.test.js:1]: TODO: Revert this file with new test cases.

**Review by KelvinOm — APPROVED** [review]

File: app/client/src/ce/utils/serviceWorkerUtils.ts [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+      this.releaseFetchMutex(request);
+    }
+  }
+
+  /**
+   * Function to get the cached response for the request
+   */
+  async getCachedResponse(request: Request) {
+    // Wait for the lock to be released
+    await this.waitForUnlock(request);
```
Comment (rajatagrawal) [inline app/client/src/ce/utils/serviceWorkerUtils.ts:228]: waitForUnlock returns a null value in some cases. Is it needed to handle that scenario ?

File: app/client/src/ce/utils/serviceWorkerUtils.ts [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  // Function to get the request key
+  getRequestKey = (request: Request) => {
+    const headersKey = Array.from(request.headers.entries())
+      .map(([key, value]) => `${key}:${value}`)
+      .join(",");
+    return `${request.method}:${request.url}:${headersKey}`;
+  };
+
+  // Function to acquire the fetch mutex for a request
+  aqcuireFetchMutex = async (request: Request) => {
```
Comment (rajatagrawal) [inline app/client/src/ce/utils/serviceWorkerUtils.ts:166]: we could consider moving mutex functions into a separate file to improve readability.

Comment (rajatagrawal) [inline app/client/src/ce/utils/serviceWorkerUtils.ts:166]: we could consider moving mutex functions into a separate file to improve readability.

Comment (rajatagrawal) [reply app/client/src/ce/utils/serviceWorkerUtils.ts:166]: and have separate jest test cases for the mutex file

File: app/client/src/ce/utils/serviceWorkerUtils.ts [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  applicationProps: TApplicationParams,
+): Request[] => {
+  return [];
+};
+
+/**
+ * Cache strategy for Appsmith API
+ */
+export class PrefetchApiCacheStrategy {
```
Comment (rajatagrawal) [inline app/client/src/ce/utils/serviceWorkerUtils.ts:149]: nit: This can be renamed to PrefetchApiCachingService

Comment (rajatagrawal) [reply app/client/src/ce/utils/serviceWorkerUtils.ts:149]: or PrefetchApiService

File: app/client/src/serviceWorker.ts [hunk 4]
```diff
@@ -30,33 +34,52 @@ const regexMap = {
...
+    if (consolidatedApiPrefetchRequest) {
+      prefetchApiCacheStrategy
+        .cacheApi(consolidatedApiPrefetchRequest)
+        .catch(() => {
+          // Silently fail
+        });
+    }
+
+    // Prefetch the module API requests
+    const moduleApiPrefetchRequests =
```
Comment (rajatagrawal) [inline app/client/src/serviceWorker.ts:75]: Can this logic be extracted into a generic functions that returns an array of all prefetch requests and fetches them ? Consolidated apis and prefetch apis.

File: app/client/src/ce/utils/serviceWorkerUtils.ts (hunk 2 above)
Comment (dvj1988) [reply app/client/src/ce/utils/serviceWorkerUtils.ts:166]: The mutex map is a state specific to this class. The methods `aqcuireFetchMutex`, `waitForUnlock` and `releaseFetchMutex` are member fns which are member fns which access or mutate this state. I would suggest keeping this as part of this class itself.

File: app/client/src/ce/utils/serviceWorkerUtils.ts (hunk 1 above)
Comment (dvj1988) [reply app/client/src/ce/utils/serviceWorkerUtils.ts:228]: if there are no prefetched requests for a certain request, then the `waitForUnlock` will return undefined. The fn `waitForUnlock` is async and can be awaited in these cases as well. Therefore no changes are required here.

**Review by rajatagrawal — APPROVED** [review]

**Pull request merged**